Week06

safiweb opened the pull request:
# Adds unit tests for a simple product factory application example

This PR adds unit tests for a simple product factory application example.  There is a _Manager_ that receives orders for _Products_. The _Manager_ assigns those products to an _Employee_ to build. When the _Employee_ finishes the _Product_ they send it to the _Manager_ who returns a _CompletedProduct_ back to the customer.

The _Manager_ makes use of goroutines and channels to carry out the various tasks. The basic sends and receives operations on the channels are blocking. The channels are also unbuffered.

The unit tests added are as follows:

**_TestCompletedProduct_IsValid_** - checks if CompleteProduct is built when assigned to an Employee by the Manager.

**_TestEmployee_IsValid_** - checks if an Employee is valid to be assigned Products by the Manager or a job.

**_TestProduct_BuiltBy_**  -  checks if the employee that built a Product is valid or not

**_TestProduct_Build_** -  checks if a product is built by  a given employee or else there is an error in the operation

**_TestProduct_IsBuilt_** - checks if a product is valid or was built correctly by the assigned Employee.

**_TestManager_Start_** - checks if employees have been created for the given count and if they have started listening for the jobs.

**_TestManager_Assign_** - checks if the Manager has assigned given products to employees.

**_TestManager_Complete_** -  checks if products and employees wrapped as CompletedProduct have been passed correctly to the Completed channel, ready for the Manager to dispatch to the customer.

## Changes Summary
- Added complete_product.go, errors.go, product.go, employee.go and manager.go
- Added complete_product._test.go, product_test.go, employee_test.go, and manager._test.go that have unit test functions
- Miscellaneous - formatting code

## Difficulties
- Initializing the manager by NewManager() for certain subtest cases running goroutines was bringing out issues i.e RACE DATA WARNING. This was seen in TestManager_Start and TestManager_Complete unit tests e.g

```
       testcases := []struct {
		name     string
		employee int
                m *Manager
		product  *Product
		err      error
	}{}
```

The remedy was to initialize the manager way before the test cases.

- Still reading further regarding Channels, the practical explanation is easy to grasp than the coding part :)

## Surprises

- RACE DATA WARNING. as mentioned above

- Sometimes `go test -v -cover` brings different results from running `go test -cover -v -race ./...`

safiweb committed:
Week06 Initial Setup

safiweb committed:
Unit Tests

Unit tests for:
- Employee
- Manager
- Products

markbates commented on week06/completed_product_test.go:
```diff
@@ -0,0 +1,71 @@
+	t.Parallel()
+
+	//Build a product of 1 Quantity for 1 Employee
+	myProduct := Product{Quantity: 1}
+	myProduct.Build(1)
+
+	//Build a product of 0 Quantity for 1 Employee
+	myProductNoQuantity := Product{}
+	myProductNoQuantity.Build(1)
+
+	//Different Complete Products for different scenarios
+	emptyCompleteProduct := CompletedProduct{}
+	completeProductNoEmployee := CompletedProduct{Product: myProduct, Employee: Employee(0)}
+	completeProductNotBuild := CompletedProduct{Product: Product{Quantity: 1}, Employee: Employee(1)}
+	completeProductNoQuantity := CompletedProduct{Product: myProductNoQuantity, Employee: Employee(1)}
+	completeProductFull := CompletedProduct{Product: myProduct, Employee: Employee(1)}
```
These variables are far too long for idiomatic Go. Try shortening them.

In addition, you are testing too much. The `CompletedProduct` validates against `Product.IsBuilt` so as long as that returns `false`, you're good. You don't need to test additional varieties of that failure here, such as an invalid quantity on the `Product`. Those scenarios belong with the unit test for `Product.IsBuilt`.

```suggestion
	prod := Product{Quantity: 1}
	prod.Build(1)
	
	e := Employee(1)

	//Different Complete Products for different scenarios
	
	empty := CompletedProduct{}
	noEmployee := CompletedProduct{Product: prod}
	notBuilt := CompletedProduct{Product: Product{Quantity: 1}, Employee: e}
	good := CompletedProduct{Product: myProduct, Employee: e}
```

markbates commented on week06/manager_test.go:
```diff
@@ -0,0 +1,211 @@
+	t.Parallel()
+
+	//create new manager
+	manager := NewManager()
```
You should probably move this into the sub-test itself. You are starting and stopping a resource shared between your sub-tests.


It is always import to clean up your resources. Currently, if your test fails it will leak resources because the `m.Stop` function was never called to clean them up. You should immediately defer a call to `m.Stop` after creating your `Manager`.

```suggestion
	manager := NewManager()
	defer manager.Stop()
```

markbates commented on week06/manager_test.go:
```diff
@@ -0,0 +1,211 @@
+	m := NewManager()
+	m.Stop()
+
+	fakeProduct := Product{}
+
+	readyProduct := Product{Quantity: 1}
+	readyProduct.Build(0)
+
+	fullProduct := Product{Quantity: 1}
+	fullProduct.Build(1)
```
```suggestion
	fake := Product{}

	ready := Product{Quantity: 1}
	ready.Build(0)

	full := Product{Quantity: 1}
	full.Build(1)
```

markbates commented on week06/manager_test.go:
```diff
@@ -0,0 +1,211 @@
+func TestManager_Complete(t *testing.T) {
+	t.Parallel()
+
+	manager := NewManager()
```
Move this into the sub-tests and clean up resources.
```suggestion
	manager := NewManager()
	defer manager.Stop()
```

markbates reviewed (commented):
Excellent work!

Please see my comments about cleaning up resources and variable names.

safiweb committed:
Code Refactoring

- Resource Management by deferring Manger
- Variable shortening to be Go idiomatic
- Testing work() as its allowed to be tested.

safiweb committed:
Code cleanup

Removing commented code

safiweb committed:
Code cleanup - resources

safiweb committed:
Resources clean up

safiweb committed:
Fixing resources leakage
